src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,375 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+#nullable enable
+
+namespace System.Reflection.Metadata
+{
+    internal static class TypeNameParserHelpers
+    {
+        internal const sbyte SZArray = -1;
+        internal const sbyte Pointer = -2;
+        internal const sbyte ByRef = -3;
+        private const char EscapeCharacter = '\\';
+#if NET8_0_OR_GREATER
+        private static readonly SearchValues<char> _endOfFullTypeNameDelimitersSearchValues = SearchValues.Create("[]&*,+\\");
+#endif
+
+        internal static string GetGenericTypeFullName(ReadOnlySpan<char> fullTypeName, IReadOnlyList<TypeName> genericArgs)
+        {
+            ValueStringBuilder result = new(stackalloc char[128]);
+            result.Append(fullTypeName);
+
+            result.Append('[');
+            foreach (TypeName genericArg in genericArgs)
+            {
+                result.Append('[');
+                result.Append(genericArg.AssemblyQualifiedName);
+                result.Append(']');
+                result.Append(',');
+            }
+            result[result.Length - 1] = ']'; // replace ',' with ']'
```

Comment:
What if genericArgs is empty?

Reply:
It can not be empty, I am going to add a debug assert to make sure it never changes.